compiler/qsc/src/interpret/stateful.rs

```diff
@@ -439,6 +437,122 @@ impl Interpreter {
+        expr: &str,
+        shots: u32,
+    ) -> Result<Vec<LineResult>, Vec<LineError>> {
+        let mut fragments = self
+            .compiler
+            .compile_expr("<entry>", expr)
+            .map_err(|errors| {
+                errors
+                    .into_iter()
+                    .map(|error| {
+                        LineError(WithSource::from_map(
+                            self.compiler.source_map(),
+                            error.into(),
+                        ))
+                    })
+                    .collect::<Vec<_>>()
+            })?;
+
+        let pass_errors = fragments
+            .iter_mut()
+            .flat_map(|fragment| {
+                self.passes
+                    .run(self.store.core(), self.compiler.assigner_mut(), fragment)
+            })
+            .collect::<Vec<_>>();
+        if !pass_errors.is_empty() {
+            return Err(pass_errors
+                .into_iter()
+                .map(|error| {
+                    LineError(WithSource::from_map(
+                        self.compiler.source_map(),
+                        error.into(),
+                    ))
+                })
+                .collect());
+        }
+
+        let mut stmt_ids = Vec::new();
+        for fragment in fragments {
+            match fragment {
+                Fragment::Item(item) => match item.kind {
+                    qsc_hir::hir::ItemKind::Callable(callable) => {
+                        let callable = self.lower_callable_decl(&callable);
+
+                        self.callables
+                            .insert(qsc_eval::lower::lower_local_item_id(item.id), callable);
+                    }
+                    qsc_hir::hir::ItemKind::Namespace(..) => {}
+                    qsc_hir::hir::ItemKind::Ty(..) => {
+                        self.udts
+                            .insert(qsc_eval::lower::lower_local_item_id(item.id));
+                    }
+                },
+                Fragment::Stmt(stmt) => {
+                    stmt_ids.push(self.lower_stmt(&stmt));
+                }
+            }
+        }
+
+        let stmt_id = stmt_ids
+            .pop()
+            .expect("expression should yield exactly one statement");
+        assert!(
+            stmt_ids.is_empty(),
+            "expression should yield exactly one statement"
+        );
+
+        let globals = Lookup {
+            fir_store: &self.fir_store,
+            package: self.package,
+            udts: &self.udts,
+            callables: &self.callables,
+        };
+
+        let mut results: Vec<LineResult> = Vec::new();
+        for _i in 0..shots {
+            results.push(
+                match eval_stmt(
+                    stmt_id,
+                    &globals,
+                    &mut Env::with_empty_scope(),
+                    &mut SparseSim::new(),
+                    self.package,
+                    receiver,
+                ) {
```

Comment:
I'm still torn on the environment part of it... I think what you have here is probably the right place to start, but putting an alternative here just for future consideration. If we want to allow the user to provide expressions that make use of variables they have defined in the interactive environment, we could add `Clone` derives to `Scope` and `Env` structs, and then this code would work:

```suggestion
            let mut env = self.env.clone();
            results.push(
                match eval_stmt(
                    stmt_id,
                    &globals,
                    &mut env,
                    &mut SparseSim::new(),
                    self.package,
                    receiver,
                ) {
```
However, then we open up the door to reuse of globally allocated qubits, which would cause problems with the simulator (today it would panic if it recieves a qubit ID that that it has not allocated). To really make this work without the corner case crash, we would have to have an additional check of the input expression to confirm it doesn't include any free variables of qubit type.

I don't think this work would fit into the current PR, but it is something we can consider for later. Note that `qirgen` below reuses the environment rather than getting a fresh one, so it already has this corner case issue.

Reply:
Yeah, I don't think there's a quick ideal solution here, so let's reconsider after this PR. In my ideal world, I think everything - including the compilation state! - would be clone-able, and both `qirgen` and `run` would operate on a cloned state. No interference at all. But that's a big change.